### PR TITLE
Support lazy dynamic filtering in hive connector

### PR DESCRIPTION
**Pull request by raunaqmorarka**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -227,6 +234,12 @@ public TaskStatus process()
                     return TaskStatus.finished();
                 }
                 ListenableFuture<?> future;
+                if (dynamicFilteringProbeBlockingTimeoutMillis.get() > 0 && !dynamicFilter.isComplete()) {
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:237]: add `long startTime` and timeout after `dynamicFilteringProbeBlockingTimeoutMillis - startTime`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 2]
```diff
@@ -227,6 +234,12 @@ public TaskStatus process()
...
+                    future = toListenableFuture(dynamicFilter.isBlocked()
+                            .orTimeout(dynamicFilteringProbeBlockingTimeoutMillis.get(), MILLISECONDS));
+                    addExceptionCallback(future, () -> dynamicFilteringProbeBlockingTimeoutMillis.set(0));
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:240]: I think this one can be removed

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -129,7 +133,8 @@
 
     private final Table table;
     private final TupleDomain<? extends ColumnHandle> compactEffectivePredicate;
-    private final Supplier<TupleDomain<ColumnHandle>> dynamicFilterSupplier;
+    private final DynamicFilter dynamicFilter;
+    private final AtomicLong dynamicFilteringProbeBlockingTimeoutMillis;
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:137]: make this `final long dynamicFilteringProbeBlockingTimeoutMillis`

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java [hunk 4]
```diff
@@ -269,6 +281,60 @@ private static void copyTableBucketed(QueryRunner queryRunner, QualifiedObjectNa
...
+                .setIdentity(Identity.forUser("hive")
+                        .withRoles(role.map(selectedRole -> ImmutableMap.of("hive", selectedRole))
+                                .orElse(ImmutableMap.of()))
+                        .build())
+                .setCatalog(HIVE_PARTITIONED_CATALOG)
+                .setSchema(TPCH_PARTITIONED_SCHEMA)
+                .build();
+    }
+
+    public static void copyTpchTablesPartitioned(
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:296]: Add `io.prestosql.plugin.hive.HiveQueryRunner.Builder#withTpchPartitionedTables()`

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java [hunk 5]
```diff
@@ -269,6 +281,60 @@ private static void copyTableBucketed(QueryRunner queryRunner, QualifiedObjectNa
...
+            case "orders":
+            case "part":
+            case "partsupp":
+            case "supplier":
+            case "nation":
+            case "region":
+                sql = format("CREATE TABLE %s AS SELECT * FROM %s", table.getObjectName(), table);
+                break;
+            case "lineitem":
+                sql = format("CREATE TABLE %s WITH (partitioned_by=array['suppkey']) AS SELECT orderkey, partkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, shipdate, commitdate, receiptdate, shipinstruct, shipmode, comment, suppkey FROM %s", table.getObjectName(), table);
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:326]: please partition by `shiptdate` which is more natural

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java [hunk 6]
```diff
@@ -269,6 +281,60 @@ private static void copyTableBucketed(QueryRunner queryRunner, QualifiedObjectNa
...
+                break;
+            case "customer":
+                sql = format("CREATE TABLE %s WITH (partitioned_by=array['nationkey', 'mktsegment']) AS SELECT custkey, name, address, phone, acctbal, comment, nationkey, mktsegment FROM %s", table.getObjectName(), table);
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:329]: let's avoid artificial partitioning.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        assertEquals(dynamicFiltersStats.getTotalDynamicFilters(), 2L);
+        assertEquals(dynamicFiltersStats.getLazyDynamicFilters(), 2L);
+        assertEquals(dynamicFiltersStats.getReplicatedDynamicFilters(), 0L);
+        assertEquals(dynamicFiltersStats.getDynamicFiltersCompleted(), 2L);
+
+        List<DynamicFilterDomainStats> domainStats = dynamicFiltersStats.getDynamicFilterDomainStats();
+        assertEquals(domainStats.size(), 2L);
+        domainStats.forEach(stats -> {
+            assertGreaterThanOrEqual(stats.getRangeCount(), 1);
+            assertEquals(stats.getDiscreteValuesCount(), 0);
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:196]: could we add a test for `getDiscreteValuesCount` too? I think you can use `row_type` for this (I don't think row types are orderable).

**Review by sopel39 — COMMENTED** [review]
looks good

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 1 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: Changed to use a StopWatch

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java (hunk 4 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:299]: Done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java (hunk 6 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java]: Removed this

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java (hunk 5 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:330]: Partitioning on shiptdate gives ~2.5K partitions which makes the test setup very slow (triggers major GC a lot on my laptop). Partitioning suppkey gives 100 partitions which works well for testing.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 7 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:182]: RowType is orderable as long as the fields within are themselves orderable. I think we can use ColorType. 
For testing getDiscreteValuesCount specifically, would it be ok we add a test in TestDynamicFilterService instead (maybe separate PR) ? It will be much easier to do it there.

**Review by rzeyde-varada — APPROVED** [review]
LGTM :)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 8]
```diff
@@ -227,6 +236,17 @@ public TaskStatus process()
                     return TaskStatus.finished();
                 }
                 ListenableFuture<?> future;
+                /*
+                 * Block until one of below conditions is met:
+                 * Arrival of first non "all" filter
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:241]: I don't think it's needed. We short-circut on all predicate.

Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:241]: put each condition in bulletpoints

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 9]
```diff
@@ -227,6 +236,17 @@ public TaskStatus process()
...
+                 * Completion of DynamicFilter
+                 * Timeout after waiting for the configured time
+                 */
+                long timeLeft = dynamicFilteringProbeBlockingTimeoutMillis - stopwatch.elapsed(MILLISECONDS);
+                if (timeLeft > 0 && !dynamicFilter.isComplete() && dynamicFilter.getCurrentPredicate().isAll()) {
+                    future = toListenableFuture(dynamicFilter.isBlocked().orTimeout(timeLeft, MILLISECONDS));
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:247]: does `orTimeout(timeLeft, MILLISECONDS)` cancel delegate future?
This is a bug in Presto. Futures returned in `DynamicFilter` are not resilient against cancellation. Could you fix it?
We probably should wrap these in `MoreFutures#unmodifiableFuture(java.util.concurrent.CompletableFuture<V>, boolean)` in Presto main

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java [hunk 10]
```diff
@@ -269,6 +284,58 @@ private static void copyTableBucketed(QueryRunner queryRunner, QualifiedObjectNa
         log.info("Imported %s rows for %s in %s", rows, table.getObjectName(), nanosSince(start).convertToMostSuccinctTimeUnit());
     }
 
+    public static Session createPartitionedSession(Optional<SelectedRole> role)
+    {
+        return testSessionBuilder()
+                .setIdentity(Identity.forUser("hive")
+                        .withRoles(role.map(selectedRole -> ImmutableMap.of("hive", selectedRole))
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:291]: move `.map` to separate line

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java (hunk 5 above)
Comment (sopel39) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java:330]: Ok. Since this is for just single test, Let's just create appropriate tables in test itself instead of using artificial keys

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java [hunk 11]
```diff
@@ -77,30 +69,4 @@ public void testJoinWithEmptyBuildSide()
         assertEquals(probeStats.getInputPositions(), 0L);
         assertEquals(probeStats.getDynamicFilterSplitsProcessed(), probeStats.getTotalDrivers());
     }
-
-    private OperatorStats searchScanFilterAndProjectOperatorStats(QueryId queryId, String tableName)
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java:81]: nit: this should go to separate commit

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 7 above)
Comment (sopel39) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:182]: > would it be ok we add a test in TestDynamicFilterService instead (maybe separate PR)

Yes, that works too

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 9 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:247]: `orTimeout` completes the underlying future exceptionally
E.g.:
```
CompletableFuture future = dynamicFilter.isBlocked();
assertFalse(future.isDone());
future.orTimeout(1, TimeUnit.MILLISECONDS);
Thread.sleep(1000);
assertFalse(future.isCancelled());
assertTrue(future.isCompletedExceptionally());
assertTrue(future.isDone());
assertFalse(dynamicFilter.isBlocked().isDone());
```
If we outright cancel the future returned by `dynamicFilter.isBlocked()` then the cancellation propagates into DynamicFilterService and prevents setting of `lazyDynamicFilters` in `addDynamicFilters`.
```
CompletableFuture future = dynamicFilter.isBlocked();
assertFalse(future.isDone());
future.cancel(false);
assertTrue(future.isCancelled());
assertTrue(dynamicFilter.isBlocked().isDone());
```
I'll make the change to warp the future in unmodifiableFuture

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java (hunk 11 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java:81]: Done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java (hunk 5 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java]: Done

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 8 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: Done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java (hunk 10 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/HiveQueryRunner.java]: Reverted these changes

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 8 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: Didn't follow this completely, my intent is to unblock the probe side scheduling as soon as we have a useful (not `all`) filter available from DynamicFilter without waiting for full completion. The future from `isBlocked` can complete multiple times with an `all` before giving a not `all` filter, so I used the condition `!dynamicFilter.isComplete() && dynamicFilter.getCurrentPredicate().isAll()` before blocking on the future from `dynamicFilter.isBlocked()`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 9 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:247]: Added a commit for wrapping in unmodifiableFuture

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java [hunk 12]
```diff
@@ -13,17 +13,9 @@
  */
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java:13]: commit message is too long

File: presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java [hunk 13]
```diff
@@ -217,7 +218,7 @@ public DynamicFilter createDynamicFilter(QueryId queryId, List<DynamicFilters.De
                     return NOT_BLOCKED;
                 }
 
-                return toCompletableFuture(whenAnyComplete(undoneFutures));
+                return unmodifiableFuture(toCompletableFuture(whenAnyComplete(undoneFutures)));
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java:221]: please add a test. Please also update `io.prestosql.sql.planner.LocalDynamicFiltersCollector.TableSpecificDynamicFilter#isBlocked` + test

File: presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java [hunk 14]
```diff
@@ -67,6 +67,7 @@
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java:67]: you could make separate PR for this fix

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 15]
```diff
@@ -227,6 +236,17 @@ public TaskStatus process()
                     return TaskStatus.finished();
                 }
                 ListenableFuture<?> future;
+                /*
+                 * Block until one of below conditions is met:
+                 *  1. Arrival of first non "all" filter
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:241]: `all` filter should trigger DF completion (and filter unblock). Remove this condition

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 16]
```diff
@@ -343,6 +346,45 @@ public void testNoHangIfPartitionIsOffline()
...
+            throws Exception
+    {
+        BackgroundHiveSplitLoader backgroundHiveSplitLoader = backgroundHiveSplitLoader(
+                new DynamicFilter() {
+                    @Override
+                    public CompletableFuture<?> isBlocked()
+                    {
+                        return CompletableFuture.runAsync(() -> {
+                            try {
+                                TimeUnit.SECONDS.sleep(30);
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:360]: make it sleep for 1H

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 17]
```diff
@@ -343,6 +346,45 @@ public void testNoHangIfPartitionIsOffline()
         assertThrows(RuntimeException.class, hiveSplitSource::isFinished);
     }
 
+    @Test(timeOut = 6_000)
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:349]: increase timeout to 30s to reduce chance for flakyness

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+    private static final Logger log = Logger.get(TestHiveDynamicPartitionPruning.class);
+    private static final String PARTITIONED_LINEITEM = "partitioned_lineitem";
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        return HiveQueryRunner.builder()
+                .setInitialTables(getTables())
+                .setHiveProperties(ImmutableMap.of("hive.dynamic-filtering-probe-blocking-timeout", "10s"))
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:65]: increase timeout for 1h to reduce risk of test flakyness

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+        DistributedQueryRunner runner = (DistributedQueryRunner) getQueryRunner();
+        ResultWithQueryId<MaterializedResult> result = runner.executeWithQueryId(
+                getSession(),
+                "SELECT * FROM partitioned_lineitem JOIN supplier ON partitioned_lineitem.suppkey = supplier.suppkey " +
+                        "AND supplier.name = 'Supplier#000000001'");
+        assertGreaterThan(result.getResult().getRowCount(), 0);
+
+        OperatorStats probeStats = searchScanFilterAndProjectOperatorStats(result.getQueryId(), "tpch:" + PARTITIONED_LINEITEM);
+        // Probe-side is partially scanned
+        assertLessThan(probeStats.getInputPositions(), countRows("lineitem"));
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:129]: question: is it a consistent same scanned row count every time?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+                getSession(),
+                "SELECT * FROM (" +
+                        "SELECT supplier.suppkey FROM " +
+                        "partitioned_lineitem JOIN tpch.tiny.supplier ON partitioned_lineitem.suppkey = supplier.suppkey AND supplier.name IN ('Supplier#000000001', 'Supplier#000000002')" +
+                        ") t JOIN tpch.tiny.partsupp ON t.suppkey = partsupp.suppkey AND partsupp.suppkey IN (2, 3)");
+        assertGreaterThan(result.getResult().getRowCount(), 0);
+
+        OperatorStats probeStats = searchScanFilterAndProjectOperatorStats(result.getQueryId(), "tpch:" + PARTITIONED_LINEITEM);
+        // Probe-side is partially scanned
+        assertLessThan(probeStats.getInputPositions(), countRows("lineitem"));
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:187]: extract `countRows("lineitem")` as static final to avoid extra queries in test

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

File: presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java (hunk 14 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java:67]: Raised https://github.com/prestosql/presto/pull/5099

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java (hunk 12 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDistributedJoinQueries.java:13]: shortened

File: presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java (hunk 13 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java]: Done in https://github.com/prestosql/presto/pull/5099

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 15 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: Removed comment

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 16 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 17 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java]: done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 18 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java]: done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 19 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java]: Yes, we use text storage format and partitioned joins so the only pruning here should come from DF and it should be consistent due to the probe blocking logic. I've updated it to assert for specific row count now.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 20 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java]: Replaced with a constant value now

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 21]
```diff
@@ -227,6 +236,16 @@ public TaskStatus process()
                     return TaskStatus.finished();
                 }
                 ListenableFuture<?> future;
+                /*
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:239]: nit: I think in methods we use:
```
// comment
//  second line
```
style of comments.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java [hunk 22]
```diff
@@ -703,6 +746,36 @@ private static void createOrcAcidFile(File file)
         return splits.build();
     }
 
+    private static BackgroundHiveSplitLoader backgroundHiveSplitLoader(
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:749]: Could you extend `io.prestosql.plugin.hive.TestBackgroundHiveSplitLoader#backgroundHiveSplitLoader(io.prestosql.plugin.hive.HdfsEnvironment, io.prestosql.spi.predicate.TupleDomain<io.prestosql.plugin.hive.HiveColumnHandle>, java.util.Optional<io.prestosql.plugin.hive.util.HiveBucketing.HiveBucketFilter>, io.prestosql.plugin.hive.metastore.Table, java.util.Optional<io.prestosql.plugin.hive.HiveBucketHandle>, java.util.Optional<org.apache.hadoop.hive.common.ValidWriteIdList>)` version of `backgroundHiveSplitLoader` with dynamic filter parameters and add an override that does provide default values for these?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        DistributedQueryRunner runner = (DistributedQueryRunner) getQueryRunner();
+        ResultWithQueryId<MaterializedResult> result = runner.executeWithQueryId(
+                getSession(),
+                "SELECT * FROM partitioned_lineitem JOIN supplier ON partitioned_lineitem.suppkey = supplier.suppkey");
+        assertGreaterThan(result.getResult().getRowCount(), 0);
+
+        OperatorStats probeStats = searchScanFilterAndProjectOperatorStats(result.getQueryId(), "tpch:" + PARTITIONED_LINEITEM);
+        // Probe-side is fully scanned
+        assertEquals(probeStats.getInputPositions(), countRows("lineitem"));
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:155]: please do something similar as `io.prestosql.plugin.memory.TestMemorySmoke#LINEITEM_COUNT` for total count of lineitem rows.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 24]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        DistributedQueryRunner runner = (DistributedQueryRunner) getQueryRunner();
+        return runner.getCoordinator()
+                .getQueryManager()
+                .getFullQueryInfo(queryId)
+                .getQueryStats()
+                .getDynamicFiltersStats();
+    }
+
+    private long countRows(String tableName)
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:214]: please add a test case where build side is too large, see: `io.prestosql.plugin.memory.TestMemorySmoke#testJoinLargeBuildSideNoDynamicFiltering`

**Review by sopel39 — APPROVED** [review]
lgtm + please add test case for large build side % minor comments

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java (hunk 22 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:749]: done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 24 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java]: Added

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 23 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java]: done

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 21 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: changed the style

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+        OperatorStats probeStats = searchScanFilterAndProjectOperatorStats(result.getQueryId(), "tpch:" + PARTITIONED_LINEITEM);
+        // Probe-side is partially scanned
+        assertBetweenInclusive(probeStats.getInputPositions(), 558L, LINEITEM_COUNT - 1);
+        assertEquals(probeStats.getDynamicFilterSplitsProcessed(), 0L);
+
+        DynamicFiltersStats dynamicFiltersStats = getDynamicFilteringStats(result.getQueryId());
+        assertEquals(dynamicFiltersStats.getTotalDynamicFilters(), 2L);
+        assertEquals(dynamicFiltersStats.getLazyDynamicFilters(), 2L);
+        assertEquals(dynamicFiltersStats.getReplicatedDynamicFilters(), 0L);
+        assertBetweenInclusive(dynamicFiltersStats.getDynamicFiltersCompleted(), 1, 2);
```
Comment (raunaqmorarka) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java:221]: @sopel39 FYI I tweaked this test a bit to assert on completion of only one filter. I was seeing some flakiness without this because the probe side gets unblocked after receiving one filter and the query can sometimes finish before DynamicFilterService is able to collect the 2nd filter. If I set `experimental.dynamic-filtering-refresh-interval` to a very low value like `20ms` then the flakiness goes away.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 26]
```diff
@@ -227,6 +236,14 @@ public TaskStatus process()
                     return TaskStatus.finished();
                 }
                 ListenableFuture<?> future;
+                // Block until one of below conditions is met:
+                // 1. Completion of DynamicFilter
+                // 2. Timeout after waiting for the configured time
+                long timeLeft = dynamicFilteringProbeBlockingTimeoutMillis - stopwatch.elapsed(MILLISECONDS);
+                if (timeLeft > 0 && !dynamicFilter.isComplete() && dynamicFilter.getCurrentPredicate().isAll()) {
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:243]: please remove `dynamicFilter.getCurrentPredicate().isAll()` predicate from if

Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java (hunk 25 above)
Comment (raunaqmorarka) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveDynamicPartitionPruning.java]: After changing to wait on full DF, the above tweak is no longer needed

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
merged, thanks!
